src/miral/wayland_extensions.cpp

```diff
@@ -40,6 +42,52 @@ auto vec2set(std::vector<std::string> vec) -> std::set<std::string>
 {
     return std::set<std::string>{vec.begin(), vec.end()};
 }
+
+struct StaticExtensionTracker
+{
+    static void add_server_extension(mir::Server* server, void* extension)
+    {
+        std::lock_guard<decltype(mutex)> lock{mutex};
+
+        for (auto const& x : extensions)
+        {
+            if (x.server == server)
+            {
+                if (x.extension == extension)
+                {
+                    return;
+                }
+                else
+                {
+                    throw mir::AbnormalExit{"ERROR: multiple miral::WaylandExtensions registered with server"};
+                }
+            }
+        }
+        extensions.push_back({server, extension});
+    }
+
+    static void remove_extension(void* extension)
+    {
+        std::lock_guard<decltype(mutex)> lock{mutex};
+
+        for (auto x = begin(extensions); x != end(extensions); ++x)
+        {
+            if (x->extension == extension)
+            {
+                extensions.erase(x);
```

Comment:
I was unaware this was safe. I assume you know what you're doing, but perhaps a `std::remove_if` would be more obviously safe?

Reply:
Why wouldn't it be safe? There is no iterator being invalidated or incremented past `end(extensions)`.